### PR TITLE
[AN-226] Enable submitting workflows to Cromwell's GCP Batch backend

### PR DESCRIPTION
**Pull request by sam-schu**
Ticket: [AN-226](https://broadworkbench.atlassian.net/browse/AN-226)

- Updated the `WorkflowSubmissionActor` to submit workflows to Cromwell's GCP Batch backend, rather than to the PAPI backend, if the `UseCromwellGcpBatchBackendSetting` workspace setting is enabled.
  - This workspace setting was added in [AN-225](https://broadworkbench.atlassian.net/browse/AN-225) ([PR here](https://github.com/broadinstitute/rawls/pull/3136)). It is currently exposed only by the workspace settings API, but will later be added to the Terra UI workspace settings modal under a feature flag.
  - Uses a new helm config value added [here](https://github.com/broadinstitute/terra-helmfile/pull/5975).
- Added the `gcp_batch_gcs_root` key, which takes the same value as the existing `jes_gcs_root`, to the execution service workflow options. This key/value pair is needed by the GCP Batch backend, and ignored by the PAPI backend.
- Updated the README and the PR template to both include all 3 places where the `rawls-model` dependency version should be updated when a new model is published.

---

**PR checklist**

- [x] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [x] Get two thumbsworth of PR review
- [x] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

[AN-226]: https://broadworkbench.atlassian.net/browse/AN-226?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[AN-225]: https://broadworkbench.atlassian.net/browse/AN-225?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala
Comment (aednichols) [inline model/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala]: I don't think "subject to change" adds any new info
```suggestion
// current possible backends are "PAPIv2-beta", "PAPIv2-CloudNAT", and "GCPBatch"
```

File: core/src/test/scala/org/broadinstitute/dsde/rawls/model/ExecutionModelSpec.scala
Comment (aednichols) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/model/ExecutionModelSpec.scala]: I am not a fan of key/value ambiguity, even for tests.

```suggestion
      gcp_batch_gcs_root = "example_gcp_batch_gcs_root",
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (aednichols) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Remove or demote below `error`.

File: README.md
Comment (aednichols) [inline README.md]: For now I recommend only checking in the content in the `Publish rawls-model` section.

I think a larger rewrite of the instructions is in order, which I intend to do in [AN-297](https://broadworkbench.atlassian.net/browse/AN-297).

In particular, I don't think it makes sense to write instructions on what **doesn't** work, people want to know what **does** work.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (jgainerdewar) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: I think this would be clearer if we did it in a single invocation (further tweaking welcome, this is just an idea):
```
useCromwellGcpBatchBackend = currentSettings.exists { setting => 
    setting match {
        case backendSetting: UseCromwellGcpBatchBackendSetting => backendSetting.config.enabled
        case _ => false
     }
}
```

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala
Comment (jgainerdewar) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: Are there any non-scalafmt or whitespace changes to this file? If not, can you back out these changes? It's better to not have files modified at all if we don't need to change them functionally.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionSpec.scala
Comment (jgainerdewar) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionSpec.scala:1211]: Nice tests!

File: local-dev/templates/rawls.conf
Comment (jgainerdewar) [inline local-dev/templates/rawls.conf:97]: Whoops, should have updated these when I switched to the new pools but I didn't know they existed!

File: model/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala
Comment (jgainerdewar) [reply model/src/main/scala/org/broadinstitute/dsde/rawls/model/ExecutionModel.scala]: Agree with Adam's suggestion and also think it would be good to note that PAPIv2-beta is not actually in use right now.

**Review by jgainerdewar — COMMENTED** [review]
This looks awesome!

File: core/src/test/scala/org/broadinstitute/dsde/rawls/model/ExecutionModelSpec.scala
Comment (sam-schu) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/model/ExecutionModelSpec.scala]: Updated!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (sam-schu) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: What do you mean by a single invocation?

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala
Comment (sam-schu) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: I'm pretty certain that I made the commit to scalafmt this file because it was giving me an error on the scalafmt PR check. I see that that action claims to check only modified files, but maybe that file was being checked because the merge commit from dev modified it. Is there a way to revert the scalafmt changes without failing the check?

File: local-dev/templates/rawls.conf
Comment (sam-schu) [inline local-dev/templates/rawls.conf:96]: @jgainerdewar I updated these pools in the local dev config to reflect the new pools you recently created because the old values were giving me trouble when trying to run Rawls locally. Do these new values seem correct? (The `exfiltrationControlled` value I was particularly unsure of.)

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (sam-schu) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: I will remove this log before merging.

Comment (sam-schu) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Please see my standalone comment below.

File: README.md
Comment (sam-schu) [reply README.md]: I removed the note from lines 177-180 — I am happy to defer updates related to local workflow submission to your ticket. Did you also think I should revert my other updates other than those related to the Rawls model? I think my additional VPN reminder and additional context about using the dev database based on what we learned will help future developers not make the same mistakes that I did.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala
Comment (jgainerdewar) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: Hm, maybe reverse all of the changes so that the diff between this branch and `develop` includes no change to this file, and see how it feels? It sounds like that might be the state you started in, but worth going back to that so we can try again.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (jgainerdewar) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: I mean going from a `collectFirst` followed by a `exists`, to just an `exists`. I find it easier to read because I don't have to think about the type coming out of the `collectFirst`. I also like making it obvious that we are handling two different `false` cases - the settings exists and is set to false, or the setting doesn't exist.

File: local-dev/templates/rawls.conf
Comment (jgainerdewar) [reply local-dev/templates/rawls.conf:96]: Yup, those look correct!

File: README.md
Comment (aednichols) [reply README.md]: Please stick to updating the content in the Publish rawls-model section.

We will have horrible merge conflicts otherwise.

Comment (aednichols) [reply README.md]: I will definitely pull you in to review that PR so you can confirm I covered everything you ran into.

Comment (aednichols) [reply README.md]: https://github.com/broadinstitute/rawls/pull/3146

**Review by aednichols — APPROVED** [review]
LGTM pending resolution of README merge conflicts/duplication with https://github.com/broadinstitute/rawls/pull/3146

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:319]: there's a good comment on `ExecutionServiceWorkflowOptions` that explains why we're passing the same value as the first two arguments … perhaps add a code comment here too? I was awfully confused when I first read this code, until I drilled down into `ExecutionServiceWorkflowOptions`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: we should probably create a reusable utility method to retrieve a strongly-typed workspace setting from the settings list, a la:
```scala
  def findSetting[T](allSettings: List[WorkspaceSetting]): Option[T] = {
    allSettings.collectFirst {
      case found:T => found
    }
  }
```
if we had that, then this code could be a bit clearer, like:
```scala
useBatchSetting: Option[UseCromwellGcpBatchBackendSetting] = somwhereTheUtilityLives
        .findSetting[UseCromwellGcpBatchBackendSetting](currentSettings)
```
or
```scala
useCromwellGcpBatchBackend: Boolean = somwhereTheUtilityLives
        .findSetting[UseCromwellGcpBatchBackendSetting](currentSettings).exists(_.config.enabled)
```
you could use the same utility to calculate `separateSubmissionSetting` up at line 293, too.

I'd consider this a stretch goal/out of scope/does not block this PR. I don't want to scope-creep!

**Review by davidangb — APPROVED** [review]
I did not test it myself, but code looks good! Minor comments inline.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (calypsomatic) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:318]: is this meant to be duplicated?

File: README.md
Comment (sam-schu) [reply README.md]: That makes sense, thanks! I have reverted the other changes.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala
Comment (sam-schu) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: I was about to revert the commit that scalafmt-ed this file, but I decided to check the diff first and I couldn't see my changes anymore. It turns out that Bria must have been having the same problem, and [this PR](https://github.com/broadinstitute/rawls/pull/3107) that was merged yesterday also fixed the formatting. Does it matter now whether or not I revert the commit that made the same changes on my branch?

Comment (jgainerdewar) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorSpec.scala]: Nope, if the diff looks good you're all set!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (sam-schu) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:319]: I added a clarifying comment, thanks!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (sam-schu) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:318]: Yes! David was confused by the same thing. I added a clarifying comment as he suggested!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (sam-schu) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Thanks! @davidangb I agree that refactoring would be great for clarity and reusability in the future, but I have implemented Janet's suggestion for now.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala
Comment (jgainerdewar) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Looks like this is still here!

**Review by jgainerdewar — COMMENTED** [review]
Looks great, happy to approve once that last log line is removed!

**Review by jgainerdewar — APPROVED** [review]

**Pull request merged**
